src/guide/typescript-support.md

```diff
@@ -0,0 +1,270 @@
+      return this.message + '!'
+    }
+  }
+})
+```
+
+### Annotating Props
+
+```ts
+{ defineComponent, PropType } from 'vue'
+
+interface ComplexMessage {
+  title: string
+  okMessage: string
+  cancelMessage: string
+}
+const Component = defineComponent({
+  props: {
+    name: String,
+    success: { type: String },
+    callback: {
+      type: Function as PropType<() => void>
+    },
+    message: {
+      type: Object as PropType<ComplexMessage>,
+      required: true,
+      validator(message: ComplexMessage) {
+        return !!message.title
+      }
+    }
+  }
+})
+```
+
+If you find validator not getting type inference or member completion isn’t working, annotating the argument with the expected type may help address these problems.
+
+## Using with Composition API
+
+On `setup()` function, you don't need ass a typing to `props` parameter as it will infer types from `props` component option.
+
+```ts
+import { defineComponent } from 'vue'
+
+const Component = defineComponent({
+  props: {
+    message: {
+      type: String,
+      required: true
+    }
+  },
+
+  setup(props) {
+    /* props will be typed as
+    (parameter) props: Readonly<{
+      msg: number;
+    } & {}>
+    */
+  }
+})
+```
+
+### Typing `ref`s
+
+To type a ref, we need to specify a type when calling `ref()` method:
+
+```ts
+import { defineComponent, ref } from 'vue'
+
+const Component = defineComponent({
+  setup() {
+    const year = ref<number>(2020)
+  }
+})
+```
+
+Sometimes we may need to specify complex types for a ref's inner value. We can do that succinctly by passing a generics argument when calling ref to override the default inference:
+
+```ts
+const year = ref<string | number>('2020') // year's type: Ref<string | number>
+
+year.value = 2020 // ok!
+```
+
+If the type of the generic is unknown, it's recommended to cast `ref` to `Ref<T>`:
+
+```ts
+function useState<State extends string>(initial: State) {
+  const state = ref(initial) as Ref<State> // state.value -> State extends string
+  return state
+}
+```
+
+### Typing `reactive`
+
+Similarly to `ref`, we can define a `reactive` type on method call:
+
+```ts
+import { defineComponent, reactive } from 'vue'
+
+interface Book {
+  title: string
+  year?: number
+}
+
+export default defineComponent({
+  name: 'HelloWorld',
+  setup() {
+    const book = reactive<Book>({
+      title: 'Vue 3 Guide'
+    })
+  }
+})
+```
+
+### Typing `computed`
+
+For computed, we could specify a type with getter only, or with getter and setter:
+
+```ts
+import { defineComponent, ref, rcomputed } from "vue";
+
+export default defineComponent({
+  name: "HelloWorld",
+  setup() {
+    let count = ref<number>(0);
+
+    // read-only, will have a type of Readonly<Ref<Readonly<number>>>
+    const double = computed<number>(() => count.value * 2);
+
+    // writable, will have a type of Ref<T>
+    const triple = computed<number>({
+      get: () => count.value * 3,
+      set: (value: number) => value,
```

Comment:
the `value:number` is not required, even without `<number>` typescript is able to infer the type correct.

Ideally typescript should be able to infer 80% of the cases, we just need to give to provide some information when those types are not inferred automatically.

Reply:
I think I will add this to the beginning of the TS section 👍🏻